### DISCUSSION TITLE
How do I supply my own type_annotation_map to automap_base?

### DISCUSSION THREAD

**Discussion by mcarans** [sqlalchemy/sqlalchemy#10285] [Usage Questions] [answered]
I have a type annotation map I want used when reflecting database tables. How do I do that?

I tried the following, but the methods in `ConversionNoTZ` don't get called when loading/saving DateTime objects:
```
        engine = create_engine(db_uri, poolclass=NullPool, echo=False)
        Session = sessionmaker(bind=engine)
        Base = automap_base(declarative_base=NoTZBase)
        Base.prepare(autoload_with=engine)
```

`NoTZBase` looks like this:
```
class NoTZBase(DeclarativeBase):
    type_annotation_map = {
        datetime: ConversionNoTZ,
    }

    @declared_attr.directive
    def __tablename__(cls):
        return f"{cls.__name__.lower()}s"
```

`ConversionNoTZ` looks like this:
```
class ConversionNoTZ(TypeDecorator):
    """Convert from/to datetime with timezone from database columns that don't
    use timezone"""

    impl = DateTime
    cache_ok = True

    def process_bind_param(self, value: Optional[datetime], _):
        if value is None:
            return value
        if value.tzinfo:
            value = value.astimezone(timezone.utc)

        return value.replace(tzinfo=None)

    def process_result_value(self, value, _):
        if value is None:
            return value
        if value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)

        return value.astimezone(timezone.utc)
```

**Comment by CaselIT** [comment]
Hi,

That's expected, since when reflecting a table from a db there isn't any information that can be used with the type annotation map.

You can refer to the documentation that has an example of changing how the columns are defined https://docs.sqlalchemy.org/en/20/orm/extensions/automap.html#intercepting-column-definitions

**Reply by zzzeek** [reply]
agree, the answer to "How to use automap_base with type_annotation_map?" is....you dont !  :)

**Reply by mcarans** [reply]
@CaselIT I may not have worded my question well so I have rephrased the title. I want to supply my own type annotation map not have it inferred or reflected. 

I thought that supplying a subclass of DeclarativeBase with my own type annotation map declared in it and passing it to automap_base would enable that type annotation map to get used for any columns that match the types in the map. For example, if a reflected column was DateTime, I thought that my map would get used since it has a mapping for DateTime. However, although I see it in automap_base, it doesn't get used. Is that intended behaviour?

**Reply by CaselIT** [reply]
> I thought that supplying a subclass of DeclarativeBase with my own type annotation map declared in it and passing it to automap_base would enable that type annotation map to get used for any columns that match the types in the map. For example, if a reflected column was DateTime, I thought that my map would get used since it has a mapping for DateTime. However, although I see it in automap_base, it doesn't get used. Is that intended behaviour?

The map is `python_type -> sqlalchemy_type`.
When using automap you never have `python_type`s, you only ever have `sqlalchem_type`s, so how can automap match using the python type datetime? how does it find the python type datetime from the database table definition?

**Reply by zzzeek** [reply, marked as answer]
> I thought that supplying a subclass of DeclarativeBase with my own type annotation map declared in it and passing it to automap_base would enable that type annotation map to get used for any columns that match the types in the map. 

the type map only applies to columns that are declared explicitly using `Mapped`.   There is not currently any connection between the type map and `Column` objects that *already* have a type.   When you reflect tables, they already have a type.

> For example, if a reflected column was DateTime, I thought that my map would get used since it has a mapping for DateTime. 

it's not.

> However, although I see it in automap_base, it doesn't get used. 

correct

> Is that intended behaviour?

yes.   you can however make your own [column_reflect](https://docs.sqlalchemy.org/en/20/orm/extensions/automap.html#intercepting-column-definitions) recipe that consumes from this type map.   I'm not sure the behavior you are looking for could exactly be a built in feature since I think there are a lot of edge cases to it that would be difficult to deliver, but the general idea could be published as a recipe on the wiki perhaps.
